src/edge-addition-planarity-suite-Version_3.0.0.5/planarity.1
.TH planarity 1

.SH NAME
planarity - The Edge Addition Planarity Suite

.SH SYNOPSIS
.B planarity
.B [-q]
.B [-h|-help]
.B [-i|-info]
.B [-r]
.B [-s]
.B [-rm]
.B [-rn]

.SH DESCRIPTION
Invokes the Edge Addition Planarity Suite commandline tool.

Without argument, the tool presents a menu-driven interactive
interface.

When used in batch mode, the tool returns 0 for a planar graph, 1 for
a nonplanar graph and -1 on error.

.SH OPTIONS
.TP
.B -q
Quiet mode
.TP
.B -h, -help
Display some help
.TP
.B -i, -info
Display copyright, license and reference articles information

.SH FEATURES
Features in batch mode.
.TP
.B -r C K N
Run the C command (see below) on K random graphs with N vertices
.TP
.B -s C I O [O2]
Run the C command (see below) on a specific graph given in the I input
file, with output in the primary O file and complementary information
in the secondary O2 file.
.TP
.B -rm N O [O2]
Compute a maximal planar random graph on N vertices with output in the
primary O file and optionally the chosen graph in the O2 file.
.TP
.B -rn N O [O2]
Compute a nonplanar random graph (maximal planar and an edge) on N
vertices with output in the primary O file and optionally the chosen
graph in the O2 file.
.SH COMMANDS
Determine which algorithm implementation to run
.TP
.B -p
Planar embedding and Kuratowski subgraph isolation
.TP
.B -d
Planar graph drawing by visibility representation
.TP
.B -o
Outerplanar embedding and obstruction isolation
.TP
.B -2
Search for subgraph homeomorphic to K_{2,3}
.TP
.B -3
Search for subgraph homeomorphic to K_{3,3}
.TP
.B -4
Search for subgraph homeomorphic to K_4
.TP
.B -a
All of the above

.SH EXAMPLES
.TP
.B planarity -s -q -p infile.txt embedding.out [obstruction.out]
Process infile.txt in quiet mode (-q), putting planar embedding in 
embedding.out or (optionally) a Kuratowski subgraph in Obstruction.out
Process returns 0=planar, 1=nonplanar, -1=error
.TP
.B planarity -s -q -d infile.txt embedding.out [drawing.out]
If graph in infile.txt is planar, then put embedding in embedding.out 
and (optionally) an ASCII art drawing in drawing.out
Process returns 0=planar, 1=nonplanar, -1=error

.SH SEE ALSO
\fBplanarity -h -menu\fR for more information.


CHANGELOG.md
# Digraphs package for GAP - CHANGELOG
Copyright (C) 2014-18 by Jan De Beule, Julius Jonušas, James D. Mitchell, Michael Torpey, Wilf A. Wilson et al.

Licensing information can be found in the LICENSE file.

## Version 0.14.0 (released 23/11/2018)

This release contains bugfixes and a couple of new features.

* The operations `AsSemigroup` and `AsMonoid` for lattice and semilattice
  digraphs were added by Chris Russell in [PR
  #136](https://github.com/gap-packages/Digraphs/pull/136).
* The operation `IsDigraphColouring` was added by [James D.
  Mitchell](http://goo.gl/ZtViV6) in [PR
  #145](https://github.com/gap-packages/Digraphs/pull/145).
* In previous versions of the package, the output of `ArticulationPoints` would
  sometimes contain repeated vertices (reported by Luke Elliott in [Issue
  #140](https://github.com/gap-packages/Digraphs/issues/140), and fixed by
  [James D.  Mitchell](http://goo.gl/ZtViV6) in [PR
  #142](https://github.com/gap-packages/Digraphs/pull/142)).
* In previous versions of the package, an unexpected error was sometimes caused
  when removing an immutable set of vertices from a digraph (reported and fixed
  by [James D.  Mitchell](http://goo.gl/ZtViV6) in [PR
  #146](https://github.com/gap-packages/Digraphs/pull/146)).
* The header file `x86intrin.h` was unnecessarily being included by the kernel
  module of Digraphs (reported by [Wilf A. Wilson](http://wilf.me) in [Issue
  #147](https://github.com/gap-packages/Digraphs/issues/147), and fixed by
  [James D.  Mitchell](http://goo.gl/ZtViV6) in [PR
  #152](https://github.com/gap-packages/Digraphs/pull/152)).

[Max Horn](https://github.com/fingolfin) also contributed various compatibility
and correctness changes to the kernel module of the package.

Digraphs now requires version 4.8.1 of the [orb
package](https://gap-packages.github.io/orb), or newer.

## Version 0.13.0 (released 19/09/2018)

This release of Digraphs contains some bugfixes, along with the following new features:

* The GraphViz engine used by `Splash` is now configurable, thanks to [Markus Pfeiffer](https://www.morphism.de/~markusp).
* The properties `IsPartialOrderDigraph`, `IsPreorderDigraph`, and `IsQuasiorderDigraph` were introduced by Chris Russell, along with the following functions for visualising these kinds of digraphs:
  * `DotPartialOrderDigraph`
  * `DotPreorderDigraph`
  * `DotQuasiorderDigraph`
* The following functions for transformations and permutations were added by [James D. Mitchell](http://goo.gl/ZtViV6):
  * `IsDigraphHomomorphism`
  * `IsDigraphEpimorphism`
  * `IsDigraphMonomorphism`
  * `IsDigraphEndomorphism`
  * `IsDigraphEmbedding`
  * `IsDigraphIsomorphism`

Digraphs now requires version 4.9.0 of GAP, or newer.

## Version 0.12.2 (released 24/08/2018)

This is a minor release which contains some small adjustments to the build
system of the package.

## Version 0.12.1 (released 26/04/2018)

This is a minor release, which contains several bugfixes. The following problems
were resolved by [James D. Mitchell](http://goo.gl/ZtViV6):

* `HomomorphismDigraphFinder` sometimes failed to find a homomorphism when one existsed [[Issue #111](https://github.com/gap-packages/Digraphs/issues/111), reported by Gordon Royle];
* the documentation for `HomomorphismDigraphFinder` was
  incomplete [[Issue #112](https://github.com/gap-packages/Digraphs/issues/112)]; and
* a segmentation fault could be caused when using Digraphs with
  NautyTracesInterface, in certain cases [[Issue #114](https://github.com/gap-packages/Digraphs/issues/114)].

## Version 0.12.0 (released 31/01/2018)

This release contains bugfixes and new features. In particular, it:

* fixes [a bug in `ArticulationPoints` and `IsBiconnectedDigraph`](https://github.com/gap-packages/Digraphs/issues/102) [[Wilf A. Wilson](http://wilf.me)];
* adds the property `IsChainDigraph` [Ashley Clayton]; and
* adds the operation `IsDigraphAutomorphism` [Chris Russell].

Digraphs now requires version 4.5.1 of the IO package.

## Version 0.11.0 (released 22/11/2017)

The principal change in Digraphs version 0.11.0 is the addition of
support for computing automorphisms, canonical labellings, and isomorphisms of
digraphs with [nauty](http://pallini.di.uniroma1.it/). This
functionality requires the [NautyTracesInterface
package](https://github.com/sebasguts/NautyTracesInterface) for GAP, version 0.2
or newer. However, this is not a required package, and the default engine
remains [bliss](http://www.tcs.hut.fi/Software/bliss/). It is possible to
specify the engine that is used by Digraphs. These changes to Digraphs were made
by [James D. Mitchell](http://goo.gl/ZtViV6)].

In particular, version 0.11.0 includes the following changes:

* `BlissAutomorphismGroup` and `NautyAutomorphismGroup` are introduced.
* `DigraphCanonicalLabelling` is replaced by `BlissCanonicalLabelling` and
  `NautyCanonicalLabelling`.
* `BlissCanonicalDigraph` and `NautyCanonicalDigraph` are introduced [Chris
  Russell and [James D. Mitchell](http://goo.gl/ZtViV6)].
* `DigraphsUseNauty` and `DigraphsUseBliss` are introduced.

The property `IsHamiltonianDigraph` and the attribute `HamiltonianPath` were
added by Luke Elliott.  Additionally, this release fixes several bugs, including
one in `DigraphSymmetricClosure` and one in `CompleteDigraph`.

Digraphs now requires version 4.4.6 of the IO package.

## Version 0.10.1 (released 16/08/2017)
This is a minor release, which contains performance improvements, and fixes a
bug in `Digraph` that could cause a segmentation fault.

## Version 0.10.0 (released 20/07/2017)
This release contains new features, bugfixes, and minor improvements to the
documentation.  There is a new method for `ChromaticNumber`, which has better
performance than the previous method
[[Julius Jonusas](http://www-groups.mcs.st-and.ac.uk/~julius)
and [James D.  Mitchell](http://goo.gl/ZtViV6)].
A bug in the code for calculating homomorphisms of digraphs, which could cause

src/edge-addition-planarity-suite-Version_3.0.0.5/c/planarity.c
/*
Copyright (c) 1997-2015, John M. Boyer
All rights reserved.
See the LICENSE.TXT file for licensing information.
*/

#include "planarity.h"

void ProjectTitle()
{
	// This message is the main location of the version number.
	// The format is major.minor.maintenance.tweak
	// Major is for an overhaul (e.g. many features, data structure change)
	// Minor is for feature addition, e.g. a new algorithm implementation added
	// Maintenance is for functional revision, e.g. bug fix to algorithm implementation
	// Tweak is for a non-functional revision, e.g. change of build scripts or testing code

	// If the version here is increased, also increase it in configure.ac
	// Furthermore, a change of Major, Minor or Maintenance here should cause a change
	// of Current, Revision and/or Age as documented in configure.ac

    Message("\n=================================================="
            "\nThe Edge Addition Planarity Suite version 3.0.0.5"
            "\nCopyright (c) 1997-2016 by John M. Boyer"
    		"\nContact info: jboyer at acm.org"
            "\n=================================================="
            "\n");
}

/****************************************************************************
 MAIN
 ****************************************************************************/

int main(int argc, char *argv[])
{
	int retVal=0;

	if (argc <= 1)
		retVal = menu();

	else if (argv[1][0] == '-')
		retVal = commandLine(argc, argv);

	else
		retVal = legacyCommandLine(argc, argv);

	// Close the log file if logging
	gp_Log(NULL);

	return retVal;
}

/****************************************************************************
 helpMessage()
 ****************************************************************************/

int helpMessage(char *param)
{
	char *commandStr =
    	"C = command (algorithm implementation to run)\n"
    	"    -p = Planar embedding and Kuratowski subgraph isolation\n"
        "    -d = Planar graph drawing by visibility representation\n"
        "    -o = Outerplanar embedding and obstruction isolation\n"
        "    -2 = Search for subgraph homeomorphic to K_{2,3}\n"
        "    -3 = Search for subgraph homeomorphic to K_{3,3}\n"
        "    -4 = Search for subgraph homeomorphic to K_4\n"
		"    -a = All of the above\n"
    	"\n";

	ProjectTitle();

	if (param == NULL)
	{
	    Message(
            "'planarity': if no command-line, then menu-driven\n"
            "'planarity (-h|-help)': this message\n"
            "'planarity (-h|-help) -gen': more help with graph generator command line\n"
            "'planarity (-h|-help) -menu': more help with menu-based command line\n"
	        "'planarity (-i|-info): copyright and license information\n"
    	    "'planarity -test [-q] [C]': runs tests (optional quiet mode, single test)\n"
	    	"\n"
	    );

	    Message(
	    	"Common usages\n"
	    	"-------------\n"
            "planarity -s -q -p infile.txt embedding.out [obstruction.out]\n"
	    	"Process infile.txt in quiet mode (-q), putting planar embedding in \n"
	    	"embedding.out or (optionally) a Kuratowski subgraph in Obstruction.out\n"
	    	"Process returns 0=planar, 1=nonplanar, -1=error\n"
	    	"\n"
            "planarity -s -q -d infile.txt embedding.out [drawing.out]\n"
            "If graph in infile.txt is planar, then put embedding in embedding.out \n"
            "and (optionally) an ASCII art drawing in drawing.out\n"
            "Process returns 0=planar, 1=nonplanar, -1=error\n"
	    );
	}

	else if (strcmp(param, "-i") == 0 || strcmp(param, "-info") == 0)
	{
	    Message(
		    "The Edge Addition Planarity Suite\n"
	    	"Copyright (c) 1997-2015, John M. Boyer\n"
		    "All rights reserved. \n"
	    	"See the LICENSE.TXT file for licensing information. \n"
            "\n"
	    	"Includes a reference implementation of the following:\n"
            "\n"
	    	"* John M. Boyer. \"Subgraph Homeomorphism via the Edge Addition Planarity \n"
	    	"  Algorithm\".  Journal of Graph Algorithms and Applications, Vol. 16, \n"
	    	"  no. 2, pp. 381-410, 2012. http://www.jgaa.info/16/268.html\n"
            "\n"
		    "* John M. Boyer. \"A New Method for Efficiently Generating Planar Graph\n"
		    "  Visibility Representations\". In P. Eades and P. Healy, editors,\n"
		    "  Proceedings of the 13th International Conference on Graph Drawing 2005,\n"
		    "  Lecture Notes Comput. Sci., Volume 3843, pp. 508-511, Springer-Verlag, 2006.\n"
            "\n"
		    "* John M. Boyer and Wendy J. Myrvold. \"On the Cutting Edge: Simplified O(n)\n"
		    "  Planarity by Edge Addition\". Journal of Graph Algorithms and Applications,\n"
		    "  Vol. 8, No. 3, pp. 241-273, 2004.\n"
		    "  http://www.jgaa.info/08/91.html\n"
            "\n"
		    "* John M. Boyer. \"Simplified O(n) Algorithms for Planar Graph Embedding,\n"
		    "  Kuratowski Subgraph Isolation, and Related Problems\". Ph.D. Dissertation,\n"
		    "  University of Victoria, 2001.\n"
            "\n"
	    );
	}

	else if (strcmp(param, "-menu") == 0)
	{
	    Message(
	    	"'planarity -r [-q] C K N': Random graphs\n"
	    	"'planarity -s [-q] C I O [O2]': Specific graph\n"
	        "'planarity -rm [-q] N O [O2]': Maximal planar random graph\n"
	        "'planarity -rn [-q] N O [O2]': Nonplanar random graph (maximal planar + edge)\n"
	        "'planarity I O [-n O2]': Legacy command-line (default -s -p)\n"
	    	"\n"
	    );

	    Message("-q is for quiet mode (no messages to stdout and stderr)\n\n");

	    Message(commandStr);

	    Message(
	    	"K = # of graphs to randomly generate\n"
	    	"N = # of vertices in each randomly generated graph\n"
	        "I = Input file (for work on a specific graph)\n"
	        "O = Primary output file\n"
	        "    For example, if C=-p then O receives the planar embedding\n"
	    	"    If C=-3, then O receives a subgraph containing a K_{3,3}\n"
	        "O2= Secondary output file\n"
	    	"    For -s, if C=-p or -o, then O2 receives the embedding obstruction\n"
	       	"    For -s, if C=-d, then O2 receives a drawing of the planar graph\n"
	    	"    For -m and -n, O2 contains the original randomly generated graph\n"
	    	"\n"
	    );

	    Message(
	        "planarity process results: 0=OK, -1=NOTOK, 1=NONEMBEDDABLE\n"
	    	"    1 result only produced by specific graph mode (-s)\n"
	        "      with command -2,-3,-4: found K_{2,3}, K_{3,3} or K_4\n"
	    	"      with command -p,-d: found planarity obstruction\n"
	    	"      with command -o: found outerplanarity obstruction\n"
	    );
	}

    FlushConsole(stdout);
    return 0;
}

/****************************************************************************
 MENU-DRIVEN PROGRAM
 ****************************************************************************/

int menu()
{
char Choice;

     do {
    	ProjectTitle();

        Message("\n"
                "P. Planar embedding and Kuratowski subgraph isolation\n"
                "D. Planar graph drawing by visibility representation\n"
                "O. Outerplanar embedding and obstruction isolation\n"
                "2. Search for subgraph homeomorphic to K_{2,3}\n"
                "3. Search for subgraph homeomorphic to K_{3,3}\n"
                "4. Search for subgraph homeomorphic to K_4\n"
        		"H. Help message for command line version\n"
                "R. Reconfigure options\n"
                "X. Exit\n"
        		"\n"
        );

        Prompt("Enter Choice: ");
        fflush(stdin);
        scanf(" %c", &Choice);
        Choice = tolower(Choice);

        if (Choice == 'h')
        	helpMessage(NULL);

        else if (Choice == 'r')
        	Reconfigure();

        else if (Choice != 'x')
        {
        	char *secondOutfile = NULL;
        	if (Choice == 'p'  || Choice == 'o' || Choice == 'd')
        		secondOutfile ="";

            switch (tolower(Mode))
            {
                case 's' : SpecificGraph(Choice, NULL, NULL, secondOutfile); break;
                case 'r' : RandomGraphs(Choice, 0, 0); break;
                case 'm' : RandomGraph(Choice, 0, 0, NULL, NULL); break;
                case 'n' : RandomGraph(Choice, 1, 0, NULL, NULL); break;
            }
        }

        if (Choice != 'r' && Choice != 'x')
        {
            Prompt("\nPress a key then hit ENTER to continue...");
            fflush(stdin);
            scanf(" %*c");
            fflush(stdin);
            Message("\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n");
            FlushConsole(stdout);
        }

     }  while (Choice != 'x');

     // Certain debuggers don't terminate correctly with pending output content
     FlushConsole(stdout);
     FlushConsole(stderr);

     return 0;
}
